# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: js/src/threading/ProtectedData.cpp
# Commit: 2fa53cd9ac4e
# Full Hash: 2fa53cd9ac4e821763dbc4402e58b58184ce230b
# Author: Jon Coppeard <jcoppeard@mozilla.com>
# Date: 2022-03-04 04:18:28
# Regressor Bug: 1643586
# File Overlap Count: 1
# Description:
#   Bug 1643586 - Part 6: Stop using JSContext for GC helper thread tasks r=jandem
#   
#   This patch moves the debug-only GC state from JSContext to JSFreeOp. Support
#   functions for this move to FreeOp.h and most RAII classes move to GCInternals.h.
#   
# ==============================================================================

diff -r 054a08d13999 -r 2fa53cd9ac4e js/src/threading/ProtectedData.cpp
--- a/js/src/threading/ProtectedData.cpp	Thu Mar 03 17:49:18 2022 +0000
+++ b/js/src/threading/ProtectedData.cpp	Thu Mar 03 17:49:18 2022 +0000
@@ -28,8 +28,7 @@
 
   if (Helper == AllowedHelperThread::GCTask ||
       Helper == AllowedHelperThread::GCTaskOrIonCompile) {
-    JSContext* cx = TlsContext.get();
-    if (cx->defaultFreeOp()->isCollecting()) {
+    if (CurrentThreadIsPerformingGC()) {
       return true;
     }
   }
@@ -102,7 +101,9 @@
 
   switch (Lock) {
     case GlobalLock::GCLock:
-      TlsContext.get()->runtime()->gc.assertCurrentThreadHasLockedGC();
+      TlsFreeOp.get()
+          ->runtimeFromAnyThread()
+          ->gc.assertCurrentThreadHasLockedGC();
       break;
     case GlobalLock::ScriptDataLock:
       TlsContext.get()->runtime()->assertCurrentThreadHasScriptDataAccess();